### PR TITLE
Extend error handing in entitlement interface

### PR DESCRIPTION
**Pull request by steffenrost**
For reasons explained in detail in this _dev.openwhisk.apache.org_ discussion thread ([Extend the SPI entitlement interface](https://lists.apache.org/thread.html/8b5ddfe5e302098370b8b4bc265e5aed5f57ba7c2c73860363a9da6a@%3Cdev.openwhisk.apache.org%3E)) we would like to change the SPI entitlement interface from Future[Boolean] to Future[Option[Boolean]]. Motivation is to signal that the entitlement request couldn't be performed at all if calling an external entitlement service failed because of a network error or unavailability of the service itself so that the caller can act more appropriate.

## Description
This code change changes the SPI entitlement interface from Future[Boolean] to Future[Option[Boolean]] and adapts the SPI entitlement implementation in _LocalEntitlementProvider_ accordingly. It also introduces a _failedToCheckAccessResourceAuthorization_ as new error response and adapts the _check_ method accordingly.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: core/controller/src/main/scala/org/apache/openwhisk/core/entitlement/Entitlement.scala [hunk 1]
```diff
@@ -295,6 +310,12 @@ protected[core] abstract class EntitlementProvider(
             if (failedResources.isEmpty) Future.successful(())
             else Future.failed(unauthorizedOn(failedResources.map(_._1)))
           })
+          .recoverWith {
+            case e =>
+              logging
+                .info(this, s"failed to check resources: ${e.getMessage}")
+              Future.failed(checkAuthorizationFailureOn(resources))
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/org/apache/openwhisk/core/entitlement/Entitlement.scala]: What do we currently respond with if an inner authorization call fails? Just a generic error?

**Comment by rabbah** [comment]
I retitled the PR to reflect the revised implementation.

**Comment by steffenrost** [comment]
@rabbah - thanks makes sense.

@markusthoemmes - thanks a lot for your review comments and discussion. Regarding your last comment (https://github.com/apache/openwhisk/pull/4709#discussion_r342501051 you are right, by just letting the inner authorization fail using the Future's failure mode we respond with a http response code 500 (internal server error) which is exactly what I wanted to achieve with the change in this pull request as it allows us to differentiate the case where the entitlement returns unauthorized (403).

I'm going to close this pull request because this change can be achieved by only adapting the iam spi entitlement provider implementation in our service. I don't see the need for doing this change in open right now. Sorry for the inconveniences.

**Comment by rabbah** [comment]
That's good to know the interface is good enough to capture your use case.
👍

**Pull request closed without merging**